### PR TITLE
mds/FSMap: allow upgrades if no up mds

### PR DESCRIPTION
**Pull request by batrick**
This is to support the fail_fs scenario for cephadm where the max_mds >= 1, !joinable, and all MDS are down.

Fixes: https://tracker.ceph.com/issues/62682

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
I've not personally tested this but I'm confident it should fix the problem.

**Review by mchangir — APPROVED** [review]
LGTM

File: src/mds/FSMap.h [hunk 1]
```diff
@@ -198,8 +198,13 @@ class Filesystem
   void print(std::ostream& out) const;
 
   bool is_upgradeable() const {
-    return (mds_map.allows_standby_replay() && mds_map.get_num_in_mds() == 0)
-       || (!mds_map.allows_standby_replay() && mds_map.get_num_in_mds() <= 1);
+    bool joinable = mds_map.joinable();
+    bool asr = mds_map.allows_standby_replay();
+    auto in_mds = mds_map.get_num_in_mds();
+    auto up_mds = mds_map.get_num_up_mds()
```
Comment (mchangir) [inline src/mds/FSMap.h:204]: nit: this line is missing a semicolon at the end

File: src/mds/FSMap.h [hunk 2]
```diff
@@ -198,8 +198,13 @@ class Filesystem
...
+    return (asr && in_mds == 0)
+           || (!asr && in_mds <= 1);
```
Comment (mchangir) [inline src/mds/FSMap.h:206]: ... and this line has an extra semicolon in the middle of a logical expression

**Comment by mchangir** [comment]
http://pulpito.front.sepia.ceph.com/mchangir-2023-09-27_06:15:50-fs:upgrade-wip-mchangir-testing-20230926.204248-distro-default-smithi/

**Comment by batrick** [comment]
Fixed!

**Comment by batrick** [comment]
Thanks for testing it with a fix Milind! I am testing a new version.

**Comment by mchangir** [comment]
[GREEN Job](http://pulpito.front.sepia.ceph.com/mchangir-2023-10-03_05:44:09-fs:upgrade-wip-mchangir-testing-20231003.084721-distro-default-smithi/)

**Comment by mchangir** [comment]
> Thanks for testing it with a fix Milind! I am testing a new version.

okay ... I just ran 1 job as a sanity test

**Comment by batrick** [comment]
> > Thanks for testing it with a fix Milind! I am testing a new version.
> 
> okay ... I just ran 1 job as a sanity test

My test run failed due to another PR. I'll merge this based on your testing.

**Pull request merged**